include/tvm/runtime/container.h

```diff
@@ -0,0 +1,259 @@
+   * \param begin The begin iterator.
+   * \param end The end iterator.
+   */
+  template <typename Iterator>
+  void assign(Iterator begin, Iterator end) {
+    size_t num_elems = std::distance(begin, end);
+    auto it = begin;
+    for (size_t i = 0; i < num_elems; ++i) {
+      void* field_ptr = AddressOf(i);
+      new (field_ptr) ElemType(*it);
+      ++it;
+    }
+    Self()->set_size(num_elems);
+  }
+
+  /*!
+   * \brief Replace the elements in the array.
+   *
+   * \param init The initializer list of elements.
+   */
+  void assign(std::initializer_list<ElemType> init) {
+    assign(init.begin(), init.end());
+  }
+
+  /*!
+   * \brief Access element at index
+   * \param idx The index of the element.
+   * \return Const reference to ElemType at the index.
+   */
+  const ElemType& operator[](size_t idx) const {
+    size_t size = Self()->size();
+    CHECK_LT(idx, size) << "Index " << idx << " out of bounds " << size << "\n";
+    return *(reinterpret_cast<ElemType*>(AddressOf(idx)));
+  }
+
+  /*!
+   * \brief Access element at index
+   * \param idx The index of the element.
+   * \return Reference to ElemType at the index.
+   */
+  ElemType& operator[](size_t idx) { return this->operator[](idx); }
+
+  /*!
+   * \brief Push a value to the end of the array.
+   *
+   * \param val The value to be added.
+   */
+  void push_back(const ElemType& val) {
+    size_t size = Self()->size();
+    CHECK_LT(size, Self()->capacity());
+    (*this)[size] = val;
+    Self()->set_size(size + 1);
+  }
+
+  /*!
+   * \brief Construct a value in the end of the array.
+   *
+   * \tparam Args Type parameters of the arguments.
+   * \param args Arguments used to construct the new value.
+   */
+  template <typename... Args>
+  void emplace_back(Args&&... args) {
+    size_t size = Self()->size();
+    CHECK_LT(size, Self()->capacity());
+    EmplaceInit(size, std::forward<Args>(args)...);
+    Self()->set_size(size + 1);
+  }
+
+  /*!
+   * \brief Destroy the Inplace Array Base object
+   */
+  ~InplaceArrayBase() {
+    if (!IsPOD()) {
+      size_t size = Self()->size();
+      for (size_t i = 0; i < size; ++i) {
+        ElemType* fp = reinterpret_cast<ElemType*>(AddressOf(i));
+        fp->ElemType::~ElemType();
+      }
+    }
+  }
+
+ protected:
+  /*!
+   * \brief Construct a value in place with the arguments.
+   *
+   * \tparam Args Type parameters of the arguments.
+   * \param idx Index of the element.
+   * \param args Arguments to construct the new value.
+   */
+  template <typename... Args>
+  void EmplaceInit(size_t idx, Args&&... args) {
+    size_t size = Self()->size();
+    CHECK_LT(size, Self()->capacity()) << "InplaceArray out of capacity\n";
+    void* field_ptr = AddressOf(size);
+    new (field_ptr) ElemType(std::forward<Args>(args)...);
+    Self()->set_size(size + 1);
+  }
+
+ private:
+  /*!
+   * \brief If the ElemType is Plain Old Data.
+   */
+  inline bool IsPOD() const {
```

Comment:
Member functions does not have to contain the inline keyword, if the implementation is in the class body